main/src/ca/uwaterloo/flix/language/phase/unification/EffUnification3.scala

```diff
@@ -0,0 +1,308 @@
+/*
+ *  Copyright 2024 Jonathan Lindegaard Starup
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+package ca.uwaterloo.flix.language.phase.unification
+
+import ca.uwaterloo.flix.api.Flix
+import ca.uwaterloo.flix.language.ast.Ast.AssocTypeConstructor
+import ca.uwaterloo.flix.language.ast.shared.Scope
+import ca.uwaterloo.flix.language.ast.{Ast, Kind, Rigidity, RigidityEnv, SourceLocation, Symbol, Type, TypeConstructor}
+import ca.uwaterloo.flix.language.phase.unification.set.Equation.Status
+import ca.uwaterloo.flix.language.phase.unification.set.SetUnification.Options
+import ca.uwaterloo.flix.language.phase.unification.set.SetFormula
+import ca.uwaterloo.flix.language.phase.unification.set.{Equation, SetSubstitution, SetUnification}
+import ca.uwaterloo.flix.util.collection.Bimap
+import ca.uwaterloo.flix.util.{InternalCompilerException, Result}
+
+import scala.io.AnsiColor
+
+object EffUnification3 {
+
+  /**
+    * Returns a most general substitution solution to `l`.
+    *
+    * @param l     the list of equations.
+    * @param scope the current region scope.
+    * @param renv  the rigidity environment.
+    * @param loc   the source location of the entire equation system, e.g. the entire function body.
+    */
+  def unifyAll(l: List[(Type, Type, SourceLocation)], scope: Scope, renv: RigidityEnv, loc: SourceLocation, opts: Options = Options.default): Result[Substitution, UnificationError] = {
+    // Compute a bi-directional map from type variables to integers.
+    implicit val bimap: Bimap[Atom, Int] = mkBidirectionalVarMap(l)
+
+    // Translate all unification problems from equations on types to equations on terms.
+    val equations = l.map {
+      case (tpe1, tpe2, loc) => toEquation(tpe1, tpe2, loc)(scope, renv, bimap)
+    }
+
+    // Compute the most-general unifier of all the equations.
+    SetUnification.solve(equations)(SetUnification.SolverListener.doNothing, opts) match {
```

Comment:
I suppose it would make sense for the solver to be a default parameter?

Also, should it not be `SolverListener.DoNothing` since its a constant?

Also, how about calling it `SolverListener.Empty`? or `Nop` etc?

Reply:
> I suppose it would make sense for the solver to be a default parameter?

not sure I understand

> Also, should it not be SolverListener.DoNothing since its a constant?

Is that how it is? i though it was lowercase in order for val/def to be an internal detail

I vote we can look at the listener in another PR